src/flamegraph/color/palette_map.rs

```diff
@@ -1,84 +1,136 @@
-            let file = BufReader::new(file);
-
-            for line in file.lines() {
-                let line = line.map_err(quick_xml::Error::Io)?;
-
-                // A line is formatted like this: NAME -> rbg(RED, GREEN, BLUE)
-                let mut words = line.split("->");
+impl PaletteMap {
+    /// Returns the color value corresponding to the given function name.
+    pub fn get(&self, func: &str) -> Option<Color> {
+        self.0.get(func).cloned()
+    }
 
-                let name = match words.next() {
-                    Some(name) => name,
-                    None => return Err(quick_xml::Error::UnexpectedToken(line)),
-                };
+    /// Inserts a function name/color pair in the map.
+    pub fn insert<S: ToString>(&mut self, func: S, color: Color) -> Option<Color> {
+        self.0.insert(func.to_string(), color)
+    }
 
-                let color = match words.next() {
-                    Some(name) => name,
-                    None => return Err(quick_xml::Error::UnexpectedToken(line)),
-                };
+    /// Provides an iterator over the elements of the map.
+    pub fn iter(&self) -> impl Iterator<Item = (&str, Color)> {
+        self.0.iter().map(|(func, color)| (func.as_str(), *color))
+    }
 
-                if words.next().is_some() {
-                    return Err(quick_xml::Error::UnexpectedToken(line));
-                }
+    /// Builds a mapping based on the inputs given by the reader.
+    ///
+    /// The reader should provide name/color pairs as text input, each pair separated by a line
+    /// separator.
+    ///
+    /// Each line should follow the format: NAME->rgb(RED, GREEN, BLUE)
+    /// where NAME is the function name, and RED, GREEN, BLUE integer values between 0 and 255
+    /// included.
+    ///
+    /// This function will return an [`std::io::Error`] if the input is not correctly formatted.
+    pub fn from_stream(reader: &mut dyn io::BufRead) -> io::Result<Self> {
+        let mut map = HashMap::default();
 
-                let rgb_color = parse_rgb_string(color)
-                    .ok_or_else(|| quick_xml::Error::UnexpectedToken(color.to_string()))?;
-                map.insert(Cow::from(name.to_string()), rgb_color);
-            }
+        for line in reader.lines() {
+            let line = line?;
+            let (name, color) = parse_line(&line)?;
+            map.insert(name.to_string(), color);
         }
 
         Ok(PaletteMap(map))
     }
 
-    pub fn save(self, file: &str) -> Result<(), io::Error> {
-        let mut file = OpenOptions::new().write(true).create(true).open(file)?;
-        let mut entries = self.0.into_iter().collect::<Vec<_>>();
+    /// Writes the palette map using the given writer.
+    /// The output content will follow the same format described in [from_stream()]
+    /// The name/color pairs will be sorted by name in lexicographic order.
+    pub fn to_stream(&self, writer: &mut dyn io::Write) -> io::Result<()> {
```

Comment:
Let's make this `to_writer` to be more consistent with other crates.

Reply:
[Done](https://github.com/jonhoo/inferno/pull/77/commits/574a80ba07903eb71fc0fded6759e77cee247afc)